middleware/markdown/process.go

```diff
@@ -0,0 +1,170 @@
+package markdown
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/russross/blackfriday"
+)
+
+const (
+	DefaultTemplate = "defaultTemplate"
+	StaticDir       = ".caddy_static"
+)
+
+// process processes the contents of a page.
+// It parses the metadata (if any) and uses the template (if found)
+func (md Markdown) process(c Config, requestPath string, b []byte) ([]byte, error) {
+	metadata, markdown, err := extractMetadata(b)
+	if err != nil {
+		return nil, err
+	}
+	// if template is not specified, check if Default template is set
+	if metadata.Template == "" {
+		if _, ok := c.Templates[DefaultTemplate]; ok {
+			metadata.Template = DefaultTemplate
+		}
+	}
+
+	// if template is set, load it
+	var tmpl []byte
+	if metadata.Template != "" {
+		if t, ok := c.Templates[metadata.Template]; ok {
+			tmpl, err = ioutil.ReadFile(t)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// process markdown
+	markdown = blackfriday.Markdown(markdown, c.Renderer, 0)
+
+	// set it as body for template
+	metadata.Variables["markdown"] = string(markdown)
+
+	return md.processTemplate(c, requestPath, tmpl, metadata)
+}
+
+// processTemplate processes a template given a requestPath,
+// template (tmpl) and metadata
+func (md Markdown) processTemplate(c Config, requestPath string, tmpl []byte, metadata Metadata) ([]byte, error) {
+	// if template is not specified,
+	// use the default template
+	if tmpl == nil {
+		tmpl = defaultTemplate(c, metadata, requestPath)
+	}
+
+	// process the template
+	b := &bytes.Buffer{}
+	t, err := template.New("").Parse(string(tmpl))
+	if err != nil {
+		return nil, err
+	}
+	if err = t.Execute(b, metadata.Variables); err != nil {
+		return nil, err
+	}
+
+	// generate static page
+	if err = md.generatePage(c, requestPath, b.Bytes()); err != nil {
+		// if static page generation fails,
+		// nothing fatal, only log the error.
+		log.Println(err)
```

Comment:
That would work - although I still wonder if just returning two error values (a fatal and non-fatal one) would be simpler, and allow the caller decide how to respond. (Sorry, I had edited my post a couple times and I know the updates don't get emailed...)